x-pack/legacy/plugins/canvas/common/lib/autocomplete.ts

```diff
@@ -0,0 +1,364 @@
+  for (const [argName, argValues] of Object.entries(fn.node.arguments)) {
+    for (let argIndex = 0; argIndex < argValues.length; argIndex++) {
+      const value = argValues[argIndex];
+      if (value.start <= position && position <= value.end) {
+        if (value.node !== null && isExpression(value)) {
+          return getFnArgAtPosition(value, position);
+        }
+        return { ast, fnIndex, argName, argIndex };
+      }
+    }
+  }
+  return { ast, fnIndex };
+}
+
+function getFnNameSuggestions(
+  specs: CanvasFunction[],
+  ast: ExpressionASTWithMeta,
+  fnIndex: number
+) {
+  // Filter the list of functions by the text at the marker
+  const { start, end, node: fn } = ast.node.chain[fnIndex];
+  const query = fn.function.replace(MARKER, '');
+  const matchingFnDefs = specs.filter(({ name }) => textMatches(name, query));
+
+  // Sort by whether or not the function expects the previous function's return type, then by
+  // whether or not the function name starts with the text at the marker, then alphabetically
+  const prevFn = ast.node.chain[fnIndex - 1];
+
+  const prevFnDef = prevFn && getByAlias(specs, prevFn.node.function);
+  const prevFnType = prevFnDef && prevFnDef.type;
+  const comparator = combinedComparator<CanvasFunction>(
+    prevFnTypeComparator(prevFnType),
+    invokeWithProp<string, 'name', CanvasFunction, number>(startsWithComparator(query), 'name'),
+    invokeWithProp<string, 'name', CanvasFunction, number>(alphanumericalComparator, 'name')
+  );
+  const fnDefs = matchingFnDefs.sort(comparator);
+
+  return fnDefs.map(fnDef => {
+    return { type: 'function', text: fnDef.name + ' ', start, end: end - MARKER.length, fnDef };
+  });
+}
+
+function getArgNameSuggestions(
+  specs: CanvasFunction[],
+  ast: ExpressionASTWithMeta,
+  fnIndex: number,
+  argName: string,
+  argIndex: number
+) {
+  // Get the list of args from the function definition
+  const fn = ast.node.chain[fnIndex].node;
+  const fnDef = getByAlias(specs, fn.function);
+  if (!fnDef) {
+    return [];
+  }
+
+  // We use the exact text instead of the value because it is always a string and might be quoted
+  const { text, start, end } = fn.arguments[argName][argIndex];
+
+  // Filter the list of args by the text at the marker
+  const query = text.replace(MARKER, '');
+  const matchingArgDefs = Object.entries<ArgValue>(fnDef.args).filter(([name]) =>
+    textMatches(name, query)
+  );
+
+  // Filter the list of args by those which aren't already present (unless they allow multi)
+  const argEntries = Object.entries(fn.arguments).map<[string, ExpressionArgASTWithMeta[]]>(
+    ([name, values]) => {
+      return [name, values.filter(value => !value.text.includes(MARKER))];
+    }
+  );
+
+  const unusedArgDefs = matchingArgDefs.filter(([_, argDef]) => {
+    if (argDef.multi) {
+      return true;
+    }
+    return !argEntries.some(([name, values]) => {
+      // @ts-ignore ArgValue doesn't have a required name type, but actual Args do
```

Comment:
This `ignore` is repeated several times... why does the value not match the type specified on line 219?  It sounds like `ArgValue` needs to have `name` be included and optional?

Reply:
All of our function definitions are typed as `ExpressionFunction`.  When they are added to the interpreters FunctionRegistry, they are run through this `Fn` [method](https://github.com/elastic/kibana/blob/master/packages/kbn-interpreter/src/common/lib/fn.js) which _slightly_ alters the type that it spits out, (one of which is that name is attached to the arg). The existing code was using that name property, which TS is complaining about.  

Now that I'm looking at this code again, I think I can change it to work without using that name, since I also have the KEY that is the same as that name but we should probably fix either that Fn method or the ExpressionFunction type in case there are other dissimilarities that pop up.